PR: https://github.com/odoo/odoo/pull/85110

From: 880954ebfc1106411b7f7a7d60aee05dfae60893
From: Gorash
Date: 2022-03-29 08:56:15

Breaking data model changes scores: del:1 + add:0, change matches:
-    _inherit = ["ir.ui.view"]

Total Changes: 70

[IMP] *: remove `_render` from `ir.ui.view` and simplify report

There were inconsistencies in the calls to `_render`.
* the view context could contain information that misled developers.
Indeed, the context and value of the view are not supposed to be found
in the rendering. Thus by calling `ir.qweb` with the name of the
template, we ensure that there is no unwanted information and in
addition the cache key is that of the name of the template which saves
a query.
* the context used for rendering was modified by a method on
`ir.ui.view`, except this is not information used by this model. There
is now a `_prepare_environment` method residing on `ir.qweb`. This
method allows to modify the value dictionary as well as the context in
which the rendering will be done. This preparation of the data as well
as my security check is done only once per rendering. This also saves
some queries
* Freeze options for rendering were inconsistent. It could be that
options on which rendering depends were not part of the cache key. Thus,
depending on the user who generated the generation of the rendering
function, there was or was not information in the template. For example
for automatic branding. This is no longer possible, because it is the
context that is used. The options serving as a cache key are only
recorded for information (for the profiling system for example). A
simplification of the `ir.qweb.field` models could be made.

The report rendering and call `ir.qweb` instead of `ir.ui.view`.

Part-of: odoo/odoo#85110

================================= pseudo patch: =================================

--- a/addons/http_routing/models/__init__.py
+++ b/addons/http_routing/models/__init__.py
@@ -2,4 +2,4 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 from . import ir_http
-from . import ir_ui_view
+from . import ir_qweb

--- a/None
+++ b/addons/http_routing/models/ir_qweb.py
@@ -0,0 +1,52 @@
+# -*- coding: ascii -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import fnmatch
+import werkzeug
+
+from odoo import models
+from odoo.http import request
+from odoo.addons.http_routing.models.ir_http import slug, unslug_url, url_for
+
+
+def keep_query(*keep_params, **additional_params):
+    """
+    Generate a query string keeping the current request querystring's parameters specified
+    in ``keep_params`` and also adds the parameters specified in ``additional_params``.
+
+    Multiple values query string params will be merged into a single one with comma seperated
+    values.
+
+    The ``keep_params`` arguments can use wildcards too, eg:
+
+        keep_query('search', 'shop_*', page=4)
+    """
+    if not keep_params and not additional_params:
+        keep_params = ('*',)
+    params = additional_params.copy()
+    qs_keys = list(request.httprequest.args) if request else []
+    for keep_param in keep_params:
+        for param in fnmatch.filter(qs_keys, keep_param):
+            if param not in additional_params and param in qs_keys:
+                params[param] = request.httprequest.args.getlist(param)
+    return werkzeug.urls.url_encode(params)
+
+
+class IrQweb(models.AbstractModel):
+    _inherit = "ir.qweb"
+
+    def _prepare_environment(self, values):
+        irQweb = super()._prepare_environment(values)
+        values['slug'] = slug
+        values['unslug_url'] = unslug_url
+        values['keep_query'] = keep_query
+
+        if (not irQweb.env.context.get('minimal_qcontext') and
+                request and request.is_frontend):
+            return irQweb._prepare_frontend_environment(values)
+
+        return irQweb
+
+    def _prepare_frontend_environment(self, values):
+        values['url_for'] = url_for
+        return self

--- a/addons/http_routing/models/ir_ui_view.py
+++ b/None
@@ -1,16 +0,0 @@
-# -*- coding: ascii -*-
-# Part of Odoo. See LICENSE file for full copyright and licensing details.
-
-from odoo import api, models
-from odoo.addons.http_routing.models.ir_http import slug, unslug_url
-
-
-class IrUiView(models.Model):
-    _inherit = ["ir.ui.view"]
-
-    @api.model
-    def _prepare_qcontext(self):
-        qcontext = super(IrUiView, self)._prepare_qcontext()
-        qcontext['slug'] = slug
-        qcontext['unslug_url'] = unslug_url
-        return qcontext
